### PR TITLE
Refactor dict_of_lists_to_lists_of_dict

### PR DESCRIPTION
**Pull request by njalerikson**
Refactor code to eliminate overly verbose and complex code. 6-12 times faster implementation.

I sought to eliminate a number of private helper functions that did not aid in simplifying the code. I also ended up adding a new function `conda_build.utils.islist` which we may find valuable to use elsewhere.

Yes, I like using list/dictionary comprehension and when formatted properly I believe they are easier to read - I totally understand how subjective this is so let me know if you disagree and want it coded differently.

I personally wish to rename `dict_of_lists_to_lists_of_dict` to `expand_variants` (and `lists_of_dict_to_dict_of_lists` to `collapse_spec`) in order to make the functions more self documented as well as making them less of a mouthful. @mingwandroid what do you think? I have no problem reverting this change.

I also moved all spec validation checks to the `validate_spec` function instead of having them spread out in various locations. This may require that `validate_spec` is rerun after combining specs in places like this:
https://github.com/conda/conda-build/blob/d8fa0b3e4edf3d98452f8fcfc381853e9955a47f/conda_build/variants.py#L517-L525

Not 100% sure some of the assumptions I made are valid, need to dig a little further into corner cases. For example, instead of joining zip_keys on some special char ("#") and then splitting the keys after the Cartesian Product, I opted to use `tuples` as the keys and at least with my test cases I get the correct result but this is likely only safe to do if all of specs keys are strings to start with.

### CODE REVIEW

**Comment by mingwandroid** [comment]
I approve very much @njalerikson! Thanks for the {clean,speed}-ups!

I would like some more reviewers on this one though if possible, @msarahan and @jjhelmus. I've always found this code a bit scary to work with personally!

**Comment by njalerikson** [comment]
I see that 2 tests are failing:

The first is failing because of the aforementioned change where all spec validation occurs separately from `dict_of_lists_to_lists_of_dict` in `validate_spec`.

The second is failing as zip_keys specify keys that are not defined in `spec`. This too is a condition that is caught in `validate_spec`.

I don't want to update these tests before I get confirmation that these overall changes are desired.

**Comment by msarahan** [comment]
Looks like a great refactor to me. Thanks for taking the time to clean it up.

File: conda_build/utils.py [hunk 1]
```diff
@@ -1175,12 +1175,74 @@ def package_has_file(package_path, file_path, refresh=False):
...
+
+    :param arg: Object to ensure is a list
+    :type arg: any
+    :return: `arg` as a `list`
+    :rtype: `list`
+    """
+    if arg is None:
+        return []
+    elif islist(arg):
+        return arg
```
Comment (njalerikson) [inline conda_build/utils.py:1189]: So I'm returning any iterable here (which is what the old `ensure_list` also did), this results in a number of places where we end up doing `list(ensure_list(obj))` to avoid generator issues and the likes.

Do we want to simply wrap this in `list` and just get it over with?

Doing so was also proposed in PR #3979.

Comment (mingwandroid) [reply conda_build/utils.py:1189]: Sounds fine to me.

**Comment by njalerikson** [comment]
Ok, I've dug further and better understand how variants are processed from config files or from the command line argument `--variants`. I haven't found anything that invalidates the assumptions I've made with this PR but I do see room for further improvement. I have a similarly large PR almost ready to go once this is accepted that reworks `conda_build.variants.combine_specs` to include much of what is in `conda_build.variants.validate_specs` to better handle clobbering/extending errors.

Are there additional changes/validations needed for this PR to move forward?

**Comment by mingwandroid** [comment]
No changes needed. Thanks a lot. Keen to see the next wave!
